src/drivers/imx/sai.c

```diff
@@ -365,8 +366,13 @@ static inline int sai_set_config(struct dai *dai, struct ipc_config_dai *common_
 	default:
 		break;
 	}
+#endif
 
+#ifndef CONFIG_IMX8ULP
 	val_cr4 |= REG_SAI_CR4_FRSZ(sai->params.tdm_slots);
+#else
+	val_cr4 |= REG_SAI_CR4_FRSZ((sai->params.tdm_slots == 1) ? 2 : sai->params.tdm_slots);
```

Comment:
Why do you add this for 8ULP?
I know that for 8ULP the TDM slots is 16, so you'll have:
`val_cr4 |= REG_SAI_CR4_FRSZ(sai->params.tdm_slots);`
the same as for non-8ULP.

Reply:
TDM slots is different slot width, slot width is 16bit, and slot is 1 for 8ulp. And when slot == 1 means only enable 1 channel but we still need two words in one frame.

Reply:
Oh, you're right, I confused `tdm_slot_width` with `tdm_slots`.

> And when slot == 1 means only enable 1 channel but we still need two words in one frame.
So, you have 2 words in frame, but only one is useful, the rest is zeroed? Or what?